app/controllers/phones_controller.rb

```diff
@@ -0,0 +1,195 @@
+    end
+
+    # Set inventory data to a new item for hash
+    def init_item(inventory)
+      item = {}
+      item[:model_name] = inventory.phone.model.name
+      item[:quantity] = inventory.quantity
+      item[:price_min] = inventory.price
+      item[:price_max] = inventory.price
+      item[:price] = "#{inventory.price}$"
+      item[:memory_min] = inventory.phone.memory.amount
+      item[:memory_max] = inventory.phone.memory.amount
+      item[:memory] = inventory.phone.memory.display_name
+      item
+    end
+
+    # Update data for existed item in hash
+    def change_item(inventory)
+      item = @warehouse[inventory.phone.model.id]
+      item[:quantity] += inventory.quantity
+
+      unless (item[:price_min] == inventory.price && item[:price_max] == inventory.price)
+        item[:price_min] = [item[:price_min], inventory.price].min
+        item[:price_max] = [item[:price_max], inventory.price].max
+        item[:price] = "#{item[:price_min]}$ - #{item[:price_max]}$"
+      end
+
+      unless (item[:memory_min] == inventory.phone.memory.amount && item[:memory_max] == inventory.phone.memory.amount)
+        item[:memory_min] = [item[:memory_min], inventory.phone.memory.amount].min
+        item[:memory_max] = [item[:memory_max], inventory.phone.memory.amount].max
+        item[:memory] = Memory.where(amount: item[:memory_min]).pick(:display_name) + " - " + Memory.where(amount: item[:memory_max]).pick(:display_name)
+      end
+      item
+    end
+
+    # Get showing data
+    def get_necessary_data
+      @models = Model.pluck(:name).to_json.html_safe
+      @body_colors = BodyColor.pluck(:name).to_json.html_safe
+      @memories = Memory.pluck(:display_name).to_json.html_safe
+    end
+
+    # Create some record if not exist
+    def create_sub_record(key, value)
+      @model = Model.where(name: value).first
+      @body_color = BodyColor.where(name: body_color_params[:name][key]).first_or_initialize
+      @memory = Memory.where(
+        amount: memory_params[:amount][key],
+        display_name: memory_params[:display_name][key]
+      ).first_or_initialize
+      @os_version = OsVersion.where(
+        major: os_version_params[:major][key],
+        minor: os_version_params[:minor][key],
+        patch: os_version_params[:patch][key]
+      ).first_or_initialize
+
+
+      @body_color.save && @memory.save && @os_version.save
+    end
+
+    # Create phone
+    def create_phone
+      @phone = Phone.where(
+        user_id: current_user.id,
+        body_color_id: @body_color.id,
+        memory_id: @memory.id,
+        os_version_id: @os_version.id,
+        model_id: @model.id
+      ).first_or_initialize
+
+      @phone.save
+    end
+
+    # Create inventory depends on
+    def create_inventory(key)
+      @inventory = Inventory.where(
+        phone_id: @phone.id,
+        price: inventory_params[:price][key]
+      ).first_or_initialize
+
+      if Inventory.exists?(id: @inventory.id)
```

Comment:
Why not `if @inventory` ? 🤔

Reply:
Cause I used `first_or_initialize` so @inventory may always return true